http/peer-id-auth.md

```diff
@@ -0,0 +1,101 @@
+# Peer ID Authentication over HTTP
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+| --------------- | ------------- | ------ | --------------- |
+| 1A              | Working Draft | Active | r0, 2023-01-23  |
+
+Authors: [@MarcoPolo]
+
+Interest Group: Same as [HTTP](README.md)
+
+## Introduction
+
+This spec defines one way of authenticating Peer IDs over HTTP using a
+challenge-response scheme.
+
+## Mutual Client and Server Peer ID Authentication
+
+1. The server initiates the authentication by responding to a request that must
+   be authenticated with the response header `WWW-Authenticate: Libp2p-Challenge
+   challenge="<base64-encoded-challenge>, Libp2p-Challenge-Server-Only"`. The
+   challenge MUST be randomly generated from server for sole purpose of
+   authenticating the client. The server SHOULD store the challenge temporarily
+   until the authentication is done. The challenge SHOULD be at least 32 bytes.
+1. The client sends a request and sets the `Authorization`
+   [header](https://www.rfc-editor.org/rfc/rfc9110.html#section-11.6.2) header
+   to the following:
+   ```
+   Libp2p-Challenge peer-id="<encoded-peer-id-bytes>",client-challenge="<base64-encoded-client-challenge>",sig="<base64-signature-bytes>"
+   ```
+   * The peer-id is encoded per the [peer-ids spec](../peer-ids/peer-ids.md).
+   * The signature is over the concatenated result of:
+   ```
+     <varint-length> + "origin=" + server-name + 
```

Comment:
How do you concat a varint and a string?

Reply:
Why not use a delimiter that is outside of the base64 range and not a valid URL character? We could use `&` for example.

Reply:
> How do you concat a varint and a string?

I can add the following section:

> The string should be utf8 encoded and the bytes should be concatenated to end end of the varint bytes.

I'll include an example as well.

Reply:
> Why not use a delimiter that is outside of the base64 range and not a valid URL character? We could use `&` for example.

Why does it matter that it's not a valid URL character? This isn't being sent over the wire. Also do you suggest a delimiter instead of a length prefix? I have a slight preference for length prefix as it's harder to mess up. You don't have to worry about properly escaping the delimiter if the delimiter shows up in the field.

Reply:
> You don't have to worry about properly escaping the delimiter if the delimiter shows up in the field.

A domain-name can only include valid URL characters, right? So we would be save with `&` as a delimiter because it cannot appear in domain-names and it is not in the base64 alphabet.

I am okay with sticking to varint prefixing aswell.